File: src/data/human/Construct_String_with_Minimum_Cost/Construct_String_with_Minimum_Cost.cpp
#include <iostream>
#include <vector>
#include <string>
#include <deque>
#include <array>
using namespace std;

constexpr int INF = 2e9;
struct TrieNode {
    int len = 0, sfx = 0, dict = 0, cost = INF;
    std::array<int, 26> next;
};

struct Trie {
    static constexpr int unassigned = 0, root = 1;
    std::vector<TrieNode> tree;

    int newNode(TrieNode node) {
        tree.push_back(node);
        return tree.size() - 1;
    }

    Trie() {
        tree.assign(2, {});
        tree[unassigned].len = -1;
        tree[unassigned].next.fill(root);
    }

    void addWord(std::string_view s, int cost) {
        int u_id = root;
        for (char c : s) {
            int x = c - 'a';
            if (tree[u_id].next[x] == unassigned) {
                tree[u_id].next[x] = newNode({.len = tree[u_id].len + 1});
            }
            u_id = tree[u_id].next[x];
        }
        tree[u_id].cost = std::min(tree[u_id].cost, cost);
    }

    void ahoCorasick() {
        std::deque<int> q{root};
        while (!q.empty()) {
            auto [u, u_id] = ref(q.front());
            q.pop_front();
            for (int x = 0; x < 26; ++x) {
                auto [v, v_id] = ref(u->next[x]);
                auto [s, s_id] = ref(tree[u->sfx].next[x]);
                if (v_id == unassigned) {
                    u->next[x] = s_id;
                } else {
                    v->sfx = s_id;
                    v->dict = s->cost != INF ? s_id : s->dict;
                    q.push_back(v_id);
                }
            }
        }
    }

    std::tuple<TrieNode*, int> ref(int id) {
        return {&tree[id], id};
    }

    TrieNode* operator[](int id) {
        return &tree[id];
    }
};

class Solution {
public:
    int minimumCost(string target, vector<string>& words, vector<int>& costs) {
        int n = target.size(), m = words.size();
        Trie trie;
        for (int i = 0; i < m; ++i) {
            trie.addWord(words[i], costs[i]);
        }
        trie.ahoCorasick();
        std::vector<int> dp(n + 1, INF);
        dp[0] = 0;
        auto u = trie[Trie::root];
        for (int i = 1; i <= n; ++i) {
            int x = target[i - 1] - 'a';
            u = trie[u->next[x]];
            auto v = u->cost != INF ? u : trie[u->dict];
            while (v->cost != INF) {
                dp[i] = std::min(dp[i], dp[i - v->len] + v->cost);
                v = trie[v->dict];
            }
        }
        return dp[n] == INF ? -1 : dp[n];
    }
};

//  Test cases for the C++ solution
#include <cassert>

// Test cases for the C++ solution
void testSolution(Solution& solution) {
    // Test case 1: Simple case with exact match and different costs
    std::string target = "abcdef";
    std::vector<std::string> words = {"abdef", "abc", "d", "def", "ef"};
    std::vector<int> costs = {100, 1, 1, 10, 5};
    int expected_result = 7;
    assert(solution.minimumCost(target, words, costs) == expected_result && "Test case 1 failed.");

    // Test case 2: Using a combination of single and multiple characters
    target = "aaaa";
    words = {"z", "zz", "zzz"};
    costs = {1, 10, 100};
    expected_result = -1;
    assert(solution.minimumCost(target, words, costs) == expected_result && "Test case 2 failed.");

    // Test case 3: No possible way to construct target
    target = "abccba";
    words = {"abc", "ab", "ccba", "c", "ba"};
    costs = {10, 1, 20, 5, 2};
    expected_result = 13;
    assert(solution.minimumCost(target, words, costs) == expected_result && "Test case 3 failed.");

    // Test case 4: Exact matches with overlapping options
    target = std::string(1000, 'a');
    words = {"a", "aa", "aaa"};
    costs = {1, 2, 3};
    expected_result = 1000;
    assert(solution.minimumCost(target, words, costs) == expected_result && "Test case 4 failed.");

    // Test case 5: Large overlap possibilities with minimum cost
    target = "xyzxyzxyz";
    words = {"xyz", "xy", "zxy", "yz", "xyzxyz"};
    costs = {5, 3, 6, 2, 8};
    expected_result = 13;
    assert(solution.minimumCost(target, words, costs) == expected_result && "Test case 5 failed.");

    // Test case 6: Only one word needed
    target = "code";
    words = {"code", "ode", "de"};
    costs = {10, 8, 7};
    expected_result = 10;  // "code" matches entirely with cost 10
    assert(solution.minimumCost(target, words, costs) == expected_result && "Test case 6 failed.");

    // Test case 7: Multiple paths but all paths are not possible
    target = "abcd";
    words = {"ab", "cd", "ef"};
    costs = {3, 4, 5};
    expected_result = 7;  // "ab" + "cd" gives the target with cost 3 + 4 = 7
    assert(solution.minimumCost(target, words, costs) == expected_result && "Test case 7 failed.");
}

int main() {
    Solution solution;
    for (int i = 0; i < 81920; ++i) {
        testSolution(solution);  
    }
    return 0;
}

File: src/data/human/Construct_String_with_Minimum_Cost/Construct_String_with_Minimum_Cost.java
class Solution {
    TrieNode root; // root of Trie structure
    Integer[] dp;  // Cached minimum costs
    int n;
    public int minimumCost(String target, String[] words, int[] costs) {
        root = new TrieNode();
        buildTrie(words, costs); // store all words in a Trie data structure with the associated costs

        n = target.length();
        dp = new Integer[n];

        int ans = topDownDP(target, 0);
        if (ans == Integer.MAX_VALUE) {
            ans = -1;
        }
        return ans;
    }

    // Return the minimum cost it takes to construct the substring `target[currIndex..]
    public int topDownDP(String target, int currIndex) {
        if (currIndex >= n) {
            return 0; // there are no more letters, so the cost is 0.
        }
        if (dp[currIndex] != null) { // if we've already calculated this index minCost
            return dp[currIndex];
        }

        // The default answer is Integer.MAX_VALUE, meaning "impossible" starting at currIndex
        int ans = Integer.MAX_VALUE;

        TrieNode currNode = root;
        for (int i = currIndex; i < n; i++) {
            char nextChar = target.charAt(i);
            if (!currNode.containsKey(nextChar)) {
                break; // no such word exists in the Trie, no point in continuing
            }
            currNode = currNode.get(nextChar); // step down the Trie

            if (currNode.endsWord) {
                int localAns = topDownDP(target, i + 1);
                // if localAns == Integer.MAX_VALUE, it's impossible to use this word as a substring of `target`
                if (localAns != Integer.MAX_VALUE) {
                    localAns += currNode.cost;
                    ans = Math.min(ans, localAns);
                }
            }
        }

        dp[currIndex] = ans;
        return ans;
    }

    private void buildTrie(String[] words, int[] costs) {
        for (int i = 0; i < words.length; i++) {
            TrieNode nextNode = root;
            for (char c : words[i].toCharArray()) {
                if (!nextNode.containsKey(c)) {
                    nextNode.put(c);
                }
                nextNode = nextNode.get(c);
            }

            if (nextNode.endsWord) {
                nextNode.cost = Math.min(nextNode.cost, costs[i]); // dumb, duplicate word condition.
            } else {
                nextNode.endsWord = true;
                nextNode.cost = costs[i];
            }
        }
    }

    class TrieNode {
        boolean endsWord = false;
        TrieNode[] childrenNodes;
        // Non-standard Trie. This Trie keeps track of costs inside the node
        int cost;
        public TrieNode() {
            childrenNodes = new TrieNode[26];
        }

        public TrieNode get(Character c) {
            return childrenNodes[c - 97];
        }

        public void put(Character c) {
            if (!containsKey(c)) {
                childrenNodes[c - 97] = new TrieNode();
            }
        }

        public boolean containsKey(Character c) {
            return childrenNodes[c - 97] != null;
        }
    }
}

public class Construct_String_with_Minimum_Cost {
public static void runTests(Solution solution) {
    // Test case 1: Simple case with valid match
    String target1 = "abcdef";
    String[] words1 = {"abdef", "abc", "d", "def", "ef"};
    int[] costs1 = {100, 1, 1, 10, 5};
    int expected_result1 = 7;
    assert solution.minimumCost(target1, words1, costs1) == expected_result1 : "Test case 1 failed. Expected " + expected_result1;

    // Test case 2: Using a combination of single and multiple characters
    String target2 = "aaaa";
    String[] words2 = {"z", "zz", "zzz"};
    int[] costs2 = {1, 10, 100};
    int expected_result2 = -1;
    assert solution.minimumCost(target2, words2, costs2) == expected_result2 : "Test case 2 failed. Expected " + expected_result2;

    // Test case 3: No possible way to construct target
    String target3 = "abccba";
    String[] words3 = {"abc", "ab", "ccba", "c", "ba"};
    int[] costs3 = {10, 1, 20, 5, 2};
    int expected_result3 = 13;
    assert solution.minimumCost(target3, words3, costs3) == expected_result3 : "Test case 3 failed. Expected " + expected_result3;

    // Test case 4: Exact matches with overlapping options
    String target4 = "a".repeat(1000);
    String[] words4 = {"a", "aa", "aaa"};
    int[] costs4 = {1, 2, 3};
    int expected_result4 = 1000;
    assert solution.minimumCost(target4, words4, costs4) == expected_result4 : "Test case 4 failed. Expected " + expected_result4;

    // Test case 5: Large overlap possibilities with minimum cost
    String target5 = "xyzxyzxyz";
    String[] words5 = {"xyz", "xy", "zxy", "yz", "xyzxyz"};
    int[] costs5 = {5, 3, 6, 2, 8};
    int expected_result5 = 13;
    assert solution.minimumCost(target5, words5, costs5) == expected_result5 : "Test case 5 failed. Expected " + expected_result5;

    // Test case 6: Only one word needed
    String target6 = "code";
    String[] words6 = {"code", "ode", "de"};
    int[] costs6 = {10, 8, 7};
    int expected_result6 = 10;
    assert solution.minimumCost(target6, words6, costs6) == expected_result6 : "Test case 6 failed. Expected " + expected_result6;

    // Test case 7: Multiple paths but all paths are not possible
    String target7 = "abcd";
    String[] words7 = {"ab", "cd", "ef"};
    int[] costs7 = {3, 4, 5};
    int expected_result7 = 7;
    assert solution.minimumCost(target7, words7, costs7) == expected_result7 : "Test case 7 failed. Expected " + expected_result7;
}


    public static void main(String[] args) {
        Solution solution = new Solution();
        for (int i = 0; i < 500000; i++) {
            runTests(solution); // Ensure your test cases are implemented inside runTests
        }
    }
}

    